// ngx-disqus/esm2015/lib/disqus.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { DisqusComponent } from './disqus.component';
import { DisqusService } from './disqus.service';
import { SHORTNAME } from './disqus.token';
import { WINDOW_PROVIDERS, WINDOW } from './window.service';
/**
 * Initialize Disqus with shortname
 * @param {?} shortname
 * @param {?} window
 * @return {?}
 */
export function DisqusFactory(shortname, window) {
    return new DisqusService(shortname, window);
}
export class DisqusModule {
    /**
     * @param {?} shortname
     * @return {?}
     */
    static forRoot(shortname) {
        return {
            ngModule: DisqusModule,
            providers: [
                WINDOW_PROVIDERS,
                { provide: SHORTNAME, useValue: shortname },
                {
                    provide: DisqusService,
                    useFactory: DisqusFactory,
                    deps: [SHORTNAME, WINDOW]
                },
            ]
        };
    }
}
DisqusModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DisqusComponent],
                exports: [DisqusComponent]
            },] },
];
