// ngx-disqus/esm2015/lib/disqus.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, Renderer2, ElementRef, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { DisqusService } from './disqus.service';
export class DisqusComponent {
    /**
     * @param {?} renderer
     * @param {?} el
     * @param {?} dService
     */
    constructor(renderer, el, dService) {
        this.renderer = renderer;
        this.el = el;
        this.dService = dService;
        /**
         * DISQUS events
         */
        this.newComment = new EventEmitter(true);
        this.ready = new EventEmitter(true);
        this.paginate = new EventEmitter(true);
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        /** Reset Disqus if any input changed */
        if (!this.dService.DISQUS) {
            this.addDisqusScript();
        }
        else {
            this.reset();
        }
    }
    /**
     * Add DISQUS script
     * @return {?}
     */
    addDisqusScript() {
        /** Set DISQUS config */
        this.dService.disqus_config = this.getConfig();
        /** @type {?} */
        const disqusScript = this.renderer.createElement('script');
        disqusScript.src = `//${this.dService.shortname}.disqus.com/embed.js`;
        disqusScript.async = true;
        disqusScript.type = 'text/javascript';
        this.renderer.setAttribute(disqusScript, 'data-timestamp', new Date().getTime().toString());
        this.renderer.appendChild(this.el.nativeElement, disqusScript);
    }
    /**
     * Reset DISQUS with the new config
     * @return {?}
     */
    reset() {
        this.dService.DISQUS.reset({
            reload: true,
            config: this.getConfig()
        });
    }
    /**
     * Create DISQUS config from inputs
     * @return {?}
     */
    getConfig() {
        /** @type {?} */
        const self = this;
        return function () {
            this.page.identifier = self.identifier;
            this.page.url = self.validateUrl(self.url);
            this.page.title = self.title;
            this.category_id = self.category;
            this.language = self.language;
            /* Available callbacks are afterRender, onInit, onNewComment, onPaginate, onReady, preData, preInit, preReset */
            this.callbacks.onNewComment = [(e) => {
                    self.newComment.emit(e);
                }];
            this.callbacks.onReady = [(e) => {
                    self.ready.emit(e);
                }];
            this.callbacks.onPaginate = [(e) => {
                    self.paginate.emit(e);
                }];
        };
    }
    /**
     * @param {?} url
     * @return {?}
     */
    validateUrl(url) {
        /** Validate URL input */
        if (url) {
            /** @type {?} */
            const r = /(http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
            if (r.test(url)) {
                return url;
            }
            else {
                console.warn('[Disqus]: Invalid URL, return undefined');
            }
        }
        /** DISQUS will fallback to "Window.location.href" when URL is undefined */
        return undefined;
    }
}
DisqusComponent.decorators = [
    { type: Component, args: [{
                selector: 'disqus',
                template: '<div id="disqus_thread"></div>',
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
DisqusComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: DisqusService }
];
DisqusComponent.propDecorators = {
    url: [{ type: Input }],
    identifier: [{ type: Input }],
    title: [{ type: Input }],
    category: [{ type: Input }],
    language: [{ type: Input }],
    newComment: [{ type: Output }],
    ready: [{ type: Output }],
    paginate: [{ type: Output }]
};
if (false) {
    /**
     * DISQUS options
     * @type {?}
     */
    DisqusComponent.prototype.url;
    /** @type {?} */
    DisqusComponent.prototype.identifier;
    /** @type {?} */
    DisqusComponent.prototype.title;
    /** @type {?} */
    DisqusComponent.prototype.category;
    /** @type {?} */
    DisqusComponent.prototype.language;
    /**
     * DISQUS events
     * @type {?}
     */
    DisqusComponent.prototype.newComment;
    /** @type {?} */
    DisqusComponent.prototype.ready;
    /** @type {?} */
    DisqusComponent.prototype.paginate;
    /** @type {?} */
    DisqusComponent.prototype.renderer;
    /** @type {?} */
    DisqusComponent.prototype.el;
    /** @type {?} */
    DisqusComponent.prototype.dService;
}
